### DISCUSSION TITLE
Reason behind Vector being a "final" class?

### DISCUSSION THREAD

**Discussion by 2jfw** [airsdk/Adobe-Runtime-Support#1672] [Q&A] [answered]
Just wondering why it is meant to not be possible to extend a `Vector` in as3 / why is this class `final`?
It is possible to extend from `Array`, so why not from `Vector`, too?

**Comment by ajwfrost** [comment, marked as answer]
Interesting question! I would assume it has to do with the slightly odd definitions used for Vector within ActionScript, so that it's actually a number of different classes depending on whether you're using a primitive vector (i.e. `Vector.<uint>` etc) or whether it's a vector of objects (e.g. `Vector.<String>`) which would use the `Vector$object` class...

See https://github.com/adobe/avmplus/blob/master/core/Vector.as if you're interested!

If you tried deriving from Vector itself, this would likely cause challenges in how the class definitions are chained up because of this fact that there are different Vector classes. And all of this then comes back to the fact that `uint` and the other primitive types are a weird combination of primitive vs object types:
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/uint.html
suggests that `uint` is a class, and you can indeed do things like `myVal.toString(16)` to get a hex version of an integer, but when you're storing a uint it is stored as a primitive type rather than a reference-counted object....

When we look at templated classes in more generic terms e.g. in C++ `template <T> class ABC {}` this is going to be an interesting challenge, to see if you could then do `var obj : ABC.<int>;`....!

**Reply by 2jfw** [reply]
@ajwfrost Very interesting! Thank you so much for the insight here, again, Andrew!
I was up to ask for the possibility of a change in order to be able to extend from `Vector` but it sounds like this the "challenges" would not do justice to the effort then. Benefit would likely also be too low.
